File: CS/Dashboard_FiscalFunctions/Fiscal Functions/GetFiscalDateFunction.cs
using DevExpress.Data.Filtering;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Dashboard_FiscalFunctions {
    abstract class GetFiscalDateFunction : ICustomFunctionOperatorBrowsable {
        protected int startDay;
        protected int startMonth;

        protected GetFiscalDateFunction(int startDay, int startMonth) {
            this.startDay = startDay;
            this.startMonth = startMonth;
        }
   
        protected abstract int GetFiscal(DateTime current);
        public int MinOperandCount => 1;

        public int MaxOperandCount => 1;

        public abstract string Description { get; }

        public FunctionCategory Category => FunctionCategory.DateTime;

        public abstract string Name { get; }

        public object Evaluate(params object[] operands) {
            return GetFiscal((DateTime) operands[0]);
        }

        public bool IsValidOperandCount(int count) {
            return count == 1;
        }

        public bool IsValidOperandType(int operandIndex, int operandCount, Type type) {
            return type == typeof(DateTime) && operandCount == 1 && operandIndex == 0;
        }

        public Type ResultType(params Type[] operands) {
            return typeof(int);
        }
    }
}

File: VB/Dashboard_FiscalFunctions/Fiscal Functions/GetFiscalDateFunction.vb
Imports DevExpress.Data.Filtering
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace Dashboard_FiscalFunctions
	Friend MustInherit Class GetFiscalDateFunction
		Implements ICustomFunctionOperatorBrowsable

		Protected startDay As Integer
		Protected startMonth As Integer

		Protected Sub New(ByVal startDay As Integer, ByVal startMonth As Integer)
			Me.startDay = startDay
			Me.startMonth = startMonth
		End Sub

		Protected MustOverride Function GetFiscal(ByVal current As Date) As Integer
		Public ReadOnly Property MinOperandCount() As Integer Implements ICustomFunctionOperatorBrowsable.MinOperandCount
			Get
				Return 1
			End Get
		End Property

		Public ReadOnly Property MaxOperandCount() As Integer Implements ICustomFunctionOperatorBrowsable.MaxOperandCount
			Get
				Return 1
			End Get
		End Property

		Public MustOverride ReadOnly Property Description() As String Implements ICustomFunctionOperatorBrowsable.Description

		Public ReadOnly Property Category() As FunctionCategory Implements ICustomFunctionOperatorBrowsable.Category
			Get
				Return FunctionCategory.DateTime
			End Get
		End Property

		Public MustOverride ReadOnly Property Name() As String Implements DevExpress.Data.Filtering.ICustomFunctionOperator.Name

		Public Function Evaluate(ParamArray ByVal operands() As Object) As Object Implements DevExpress.Data.Filtering.ICustomFunctionOperator.Evaluate
			Return GetFiscal(DirectCast(operands(0), Date))
		End Function

		Public Function IsValidOperandCount(ByVal count As Integer) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandCount
			Return count = 1
		End Function

		Public Function IsValidOperandType(ByVal operandIndex As Integer, ByVal operandCount As Integer, ByVal type As Type) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandType
			Return type Is GetType(Date) AndAlso operandCount = 1 AndAlso operandIndex = 0
		End Function

		Public Function ResultType(ParamArray ByVal operands() As Type) As Type Implements DevExpress.Data.Filtering.ICustomFunctionOperator.ResultType
			Return GetType(Integer)
		End Function
	End Class
End Namespace
